README.md
# Pretty iOS notch

Виджет позволяющий добавить под челку нужный контент 

# Размеры в пикселях

|           | longNotch | shortNotch | dynamicIsland |
|-----------|-----------|------------|---------------|
| maxHeight | 26        | 26         | 35            |
| maxWidth  | 150       | 111        | 120           |


PrettyNotch виджет должен помещаться в MaterialApp (builder) либо выше него. Работает за счет Stack

# Примечание

Данный пакета использует информацию из плагина device_info_plus, для определения типа челки 

# Пример 


![Пример в приложении бристоль](assets/example.jpg)


By @serezhia

CHANGELOG.md
## 0.0.1

* Добавил поддержку разных видов челок + базовый виджет


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:pretty_notch/pretty_notch.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        useMaterial3: true,
      ),
      home: const PrettyNotch(
        longNotch: TestNotch(),
        shortNotch: TestNotch(),
        dynamicIsland: TestNotch(),
        child: MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

class TestNotch extends StatelessWidget {
  const TestNotch({super.key});

  @override
  Widget build(BuildContext context) {
    return const Placeholder();
  }
}


lib/src/notch_util.dart
import 'dart:io';

import 'package:device_info_plus/device_info_plus.dart';
import 'package:pretty_notch/src/notch_type.dart';

class NotchUtil {
  NotchUtil._();

  static Future<IosNotchType> get notchType async {
    if (!Platform.isIOS) return IosNotchType.noting;

    final machine = (await DeviceInfoPlugin().iosInfo).utsname.machine;

    return switch (machine) {
      'iPhone16,2' => IosNotchType.dynamicIsland,
      'iPhone16,1' => IosNotchType.dynamicIsland,
      'iPhone15,5' => IosNotchType.dynamicIsland,
      'iPhone15,4' => IosNotchType.dynamicIsland,
      'iPhone15,3' => IosNotchType.dynamicIsland,
      'iPhone15,2' => IosNotchType.dynamicIsland,
      'iPhone14,8' => IosNotchType.shortNotch,
      'iPhone14,7' => IosNotchType.shortNotch,
      'iPhone14,5' => IosNotchType.shortNotch,
      'iPhone14,4' => IosNotchType.shortNotch,
      'iPhone14,3' => IosNotchType.shortNotch,
      'iPhone14,2' => IosNotchType.shortNotch,
      'iPhone13,4' => IosNotchType.longNotch,
      'iPhone13,3' => IosNotchType.longNotch,
      'iPhone13,2' => IosNotchType.longNotch,
      'iPhone13,1' => IosNotchType.longNotch,
      'iPhone12,5' => IosNotchType.longNotch,
      'iPhone12,3' => IosNotchType.longNotch,
      'iPhone12,1' => IosNotchType.longNotch,
      'iPhone11,8' => IosNotchType.longNotch,
      'iPhone11,6' => IosNotchType.longNotch,
      'iPhone11,4' => IosNotchType.longNotch,
      'iPhone11,2' => IosNotchType.longNotch,
      'iPhone10,6' => IosNotchType.longNotch,
      'iPhone10,3' => IosNotchType.longNotch,
      _ => IosNotchType.noting,
    };
  }
}


lib/src/notch_calculator.dart
import 'dart:io';

import 'package:device_info_plus/device_info_plus.dart';
import 'package:pretty_notch/src/notch_type.dart';

Future<IosNotchType> getNotchType() async {
  if (!Platform.isIOS) return IosNotchType.noting;

  final machine = (await DeviceInfoPlugin().iosInfo).utsname.machine;

  return switch (machine) {
    'iPhone16,2' => IosNotchType.dynamicIsland,
    'iPhone16,1' => IosNotchType.dynamicIsland,
    'iPhone15,5' => IosNotchType.dynamicIsland,
    'iPhone15,4' => IosNotchType.dynamicIsland,
    'iPhone15,3' => IosNotchType.dynamicIsland,
    'iPhone15,2' => IosNotchType.dynamicIsland,
    'iPhone14,8' => IosNotchType.littleNotch,
    'iPhone14,7' => IosNotchType.littleNotch,
    'iPhone14,5' => IosNotchType.littleNotch,
    'iPhone14,4' => IosNotchType.littleNotch,
    'iPhone14,3' => IosNotchType.littleNotch,
    'iPhone14,2' => IosNotchType.littleNotch,
    'iPhone13,4' => IosNotchType.longNotch,
    'iPhone13,3' => IosNotchType.longNotch,
    'iPhone13,2' => IosNotchType.longNotch,
    'iPhone13,1' => IosNotchType.longNotch,
    'iPhone12,5' => IosNotchType.longNotch,
    'iPhone12,3' => IosNotchType.longNotch,
    'iPhone12,1' => IosNotchType.longNotch,
    'iPhone11,8' => IosNotchType.longNotch,
    'iPhone11,6' => IosNotchType.longNotch,
    'iPhone11,4' => IosNotchType.longNotch,
    'iPhone11,2' => IosNotchType.longNotch,
    'iPhone10,6' => IosNotchType.longNotch,
    'iPhone10,3' => IosNotchType.longNotch,
    _ => IosNotchType.noting,
  };
}


lib/src/pretty_notch_widget.dart
import 'package:flutter/material.dart';
import 'package:pretty_notch/src/notch_util.dart';

import 'notch_type.dart';

class PrettyNotch extends StatefulWidget {
  const PrettyNotch({
    required this.child,
    super.key,
    required this.longNotch,
    required this.shortNotch,
    required this.dynamicIsland,
  });

  final Widget child;

  final Widget longNotch;

  final Widget shortNotch;

  final Widget dynamicIsland;

  @override
  State<PrettyNotch> createState() => _PrettyNotchState();
}

class _PrettyNotchState extends State<PrettyNotch> {
  late final Future<IosNotchType> notch;

  /// Используется чтобы сохранять контекст всей иерархии и не вызывать лишние
  /// ребилды если виджет будет интерактивный
  final bodyKey = GlobalKey();

  @override
  void initState() {
    notch = NotchUtil.notchType;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return OrientationBuilder(
      builder: (context, orientation) {
        if (orientation == Orientation.landscape) {
          return KeyedSubtree(key: bodyKey, child: widget.child);
        } else {
          return Stack(
            children: [
              Positioned.fill(
                child: KeyedSubtree(
                  key: bodyKey,
                  child: widget.child,
                ),
              ),
              FutureBuilder(
                future: notch,
                builder: ((context, snapshot) {
                  if (!snapshot.hasData) {
                    return const SizedBox.shrink();
                  }

                  return switch (snapshot.data!) {
                    IosNotchType.dynamicIsland => Positioned(
                        top: 12,
                        left: 0,
                        right: 0,
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            ClipRRect(
                              borderRadius: BorderRadius.circular(60),
                              child: ConstrainedBox(
                                  constraints: const BoxConstraints(
                                    maxHeight: 35,
                                    maxWidth: 120,
                                  ),
                                  child: widget.dynamicIsland),
                            ),
                          ],
                        ),
                      ),
                    IosNotchType.longNotch => Positioned(
                        top: 5,
                        left: 0,
                        right: 0,
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            ClipRRect(
                              borderRadius: BorderRadius.circular(60),
                              child: ConstrainedBox(
                                  constraints: const BoxConstraints(
                                    maxHeight: 26,
                                    maxWidth: 150,
                                  ),
                                  child: widget.shortNotch),
                            ),
                          ],
                        ),
                      ),
                    IosNotchType.shortNotch => Positioned(
                        top: 5,
                        left: 0,
                        right: 0,
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            ClipRRect(
                              borderRadius: BorderRadius.circular(60),
                              child: ConstrainedBox(
                                  constraints: const BoxConstraints(
                                    maxHeight: 26,
                                    maxWidth: 111,
                                  ),
                                  child: widget.shortNotch),
                            ),
                          ],
                        ),
                      ),
                    IosNotchType.noting => const SizedBox.shrink(),
                  };
                }),
              )
            ],
          );
        }
      },
    );
  }
}
